common/engine/keyboardprocessor/src/kmx/kmx_actions.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+  Copyright:        Copyright (C) 2003-2018 SIL International.
+  Authors:          mcdurdin
+*/
+#include <kmx/kmx_processor.h>
+
+using namespace km::kbp::kmx;
+
+void KMX_Actions::ResetQueue()
+{
+  QueueSize = 0;
+}
+
+KMX_Actions::KMX_Actions(KMX_Context *context)
+{
+  m_context = context;
+  ResetQueue();
+}
+
+KMX_BOOL KMX_Actions::QueueAction(int ItemType, KMX_DWORD dwData)
+{
+  if (QueueSize > MAXACTIONQUEUE - 1)
+  {
+    DebugLog("App::QueueAction: queue size exceeded");
+    return FALSE;
+  }
+
+  Queue[QueueSize].ItemType = ItemType;
+  Queue[QueueSize].dwData = dwData;
+
+  QueueSize++;
```

Comment:
I saw you're planning to replace Queue with a list so this may not matter...

If `QueueSize == MAXACTIONQUEUE-1`, then incrementing results in `Length() == MAXACTIONQUEUE` which could go badly.

Reply:
That should be okay -- Length can be up to the size of the queue.